File: 308 Range Sum Query 2D - Mutable.java
/*
Range Sum Query 2D - Mutable
https://www.topcoder.com/community/data-science/data-science-tutorials/binary-indexed-trees/
*/
public class NumMatrix {

    int data[][];
    int tree[][];
    int R;
    int C;

    public NumMatrix(int[][] matrix) {
        R = matrix.length;
        if(R == 0) return;
        C = matrix[0].length;
        if(C == 0) return;
        data = new int[R][C];
        // Notice the reason to use extra row and col
        tree = new int[R+1][C+1]; 
        for(int i = 0;i<R;i++)
            for(int j = 0;j<C;j++)
                update(i,j,matrix[i][j]);
    }

    public void update(int row, int col, int val) {
        if(data == null) return;
        int diff = val - data[row][col];
        data[row][col] = val;
        for(int i = row+1;i<=R;i += (i & -i))
            for(int j = col+1;j<=C;j += (j & -j))
                tree[i][j] += diff;
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        if(data == null) return 0;
        return sum(row2 + 1, col2 + 1) - sum(row2 + 1, col1) - sum(row1, col2 + 1) + sum(row1, col1);
    }
    
    private int sum(int row, int col){
        int r = 0;
        // Not i >= 0
        for(int i = row;i > 0;i -= (i & -i))
            for(int j = col;j > 0;j -= (j & -j))
                r += tree[i][j];
        return r;
    }
}

File: 308 Range Sum Query 2D - Mutable.py
'''
308. Range Sum Query 2D - Mutable
https://leetcode.com/problems/range-sum-query-2d-mutable/

Given a 2D matrix matrix, find the sum of the elements inside the rectangle defined by
its upper left corner (row1, col1) and lower right corner (row2, col2).

Range Sum Query 2D
The above rectangle (with the red border) is defined by (row1, col1) = (2, 1) and (row2, col2) = (4, 3),
which contains sum = 8.

Example:

Given matrix = [
  [3, 0, 1, 4, 2],
  [5, 6, 3, 2, 1],
  [1, 2, 0, 1, 5],
  [4, 1, 0, 1, 7],
  [1, 0, 3, 0, 5]
]

sumRegion(2, 1, 4, 3) -> 8
update(3, 2, 2)
sumRegion(2, 1, 4, 3) -> 10
'''
class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        self.isfucked = not matrix or not matrix[0]
        if self.isfucked:
            return
        
        rowCount = len(matrix)
        colCount = len(matrix[0])
        
        self.nums = [[0] * colCount for _ in range(rowCount)]
        self.bits = [[0] * (colCount+1) for _ in range(rowCount+1)]
        
        for i in range(rowCount):
            for j in range(colCount):
                self.update(i,j,matrix[i][j])
                
    def update(self, row: int, col: int, val: int) -> None:
        if self.isfucked:
            return
        
        diff = val - self.nums[row][col]
        if diff != 0:
            self.nums[row][col] = val
            
            rowCount = len(self.nums)
            colCount = len(self.nums[0])
        
            i = row + 1
            while i <= rowCount:
                j = col + 1
                while j <= colCount:
                    self.bits[i][j] += diff
                    j += (j & -j)
                i += (i & -i)

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        if self.isfucked:
            return 0
        
        return self.getSum(row2+1, col2+1) - self.getSum(row1, col2+1) - self.getSum(row2+1, col1) + self.getSum(row1, col1)
    
    def getSum(self, row: int, col: int) -> int:
        total = 0
        i = row
        while i > 0:
            j = col
            while j > 0:
                total += self.bits[i][j]
                j -= (j & -j)
            i -= (i & -i)
        return total
